api/src/controllers/copilotController.ts

```diff
@@ -2,62 +2,125 @@ import { Request, Response } from 'express';
+  completeUserPrompt += `Answer the above question from user.`;
 
-  const threadTexts: string[] = threads.map((thread, threadIndex) => {
-    const replyTexts = thread.replies
-      .map((reply, replyIndex) => `Reply ${replyIndex + 1}: ${getThreadText(reply)}`)
-      .join('\n');
-    return `Thread ${threadIndex + 1}: ${thread.title ?? ''} \n ${getThreadText(thread)} \n ${replyTexts}`;
-  });
+  if (referToDoclinThreads || referToCodeFile) {
+    completeUserPrompt += `Use the references below for your answer, only if they are relevant to the question\n`;
+    completeUserPrompt += DIVIDER;
+  }
 
-  let systemPrompt: string = `References: \n\n ${threadTexts.join('\n\n')} \n\n`;
-  systemPrompt += `---------\n\n`;
-  systemPrompt += `Answer user's prompt using the references above. Format your response in HTML`;
+  if (referToDoclinThreads) {
+    completeUserPrompt += `Previous discussion on the sytem:\n`;
+    completeUserPrompt += (await getThreadsReference(userPrompt, projectId)) ?? `No relevant discussion threads found.`;
+    completeUserPrompt += DIVIDER;
+  }
 
-  return systemPrompt;
+  if (referToCodeFile && activeEditorText) {
+    completeUserPrompt += `Currently opened code file:\n`;
+    completeUserPrompt += activeEditorText;
+  }
+
+  return completeUserPrompt;
 };
 
-const sendPromptToOpenAI = async (systemPrompt: string, userPrompt: string) => {
-  const response = await openai.chat.completions.create({
-    model: 'gpt-3.5-turbo',
-    messages: [
-      {
-        role: 'system',
-        content: systemPrompt,
-      },
-      {
-        role: 'user',
-        content: userPrompt,
-      },
-    ],
-    temperature: 1,
-    max_tokens: 4096,
-    top_p: 1,
-    frequency_penalty: 0,
-    presence_penalty: 0,
-  });
-
-  return response.choices[0].message.content;
+const getThreadsReference = async (userPrompt: string, projectId: number): Promise<string | null> => {
+  const threads: Thread[] = await ThreadRepository.searchThreads(userPrompt, projectId, 5);
+
+  if (threads.length === 0) {
+    return null;
+  }
+
+  let response = `Discussion threads reference:\n`;
+  response += threads.map(getThreadReference).join('\n');
+  response += '\n';
+
+  return response;
+};
+
+const getThreadReference = (thread: Thread, index: number, threads: Thread[]): string => {
+  let response = `Thread ${index + 1} of ${threads.length}\n`;
+  response += `${thread.title ?? ''}\n`;
+  response += `${getThreadMessageWithSnippet(thread)}\n`;
+  response += thread.replies.slice(0, 3).map(getReplyReference).join('\n');
+  return response;
+};
+
+const getReplyReference = (reply: Reply, index: number, replies: Reply[]): string => {
+  return `Reply ${index + 1} of ${replies.length}: ${getThreadMessageWithSnippet(reply)}`;
+};
+
+const getResponseFromGPT = async (userPrompt: string, previousMessages: CopilotMessage[]): Promise<string | null> => {
+  try {
+    const response = await openai.chat.completions.create({
+      model: 'gpt-4o',
```

Comment:
lets use the recently released `gpt-4o-2024-08-06` literally half the [price](https://openai.com/api/pricing/).